code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -330,62 +509,86 @@ runBroker = do
-                Nothing -> ztCliLog Warning $ "client btc: couldn't resolve peer: " <> show addr
-                Just nodeId -> do
-                    onHeartbeat nodeId
-                    clientIdM <-
-                        atomically $ Map.lookup (MsgType msgT) <$> readTVar ztMsgTypes
-                    maybe (ztCliLog Warning $ "backendToClients: couldn't find client " <>
-                                              "with this message type")
-                          (\clientId -> sendToClient clientId (nodeId, Response (MsgType msgT) msg))
-                          clientIdM
-            other -> ztCliLog Warning $ "backendToClients: wrong format: " <> show other
-
-    let subToClients = \case
-            (k:addr:content) -> resolvePeer addr >>= \case
-                Nothing -> ztCliLog Warning $ "subToClients: couldn't resolve peer: " <> show addr
-                Just nodeId -> do
-                    let k' = Subscription k
-                    onHeartbeat nodeId
-                    unless (k' == heartbeatSubscription) $ do
-                        cids <-
-                            atomically $
-                            fromMaybe mempty . Map.lookup k' <$>
-                            readTVar ztSubscriptions
-                        forM_ cids $ \clientId ->
-                          sendToClient clientId (nodeId, Update k' content)
-                        when (null cids) $
-                            -- It shouldn't be alright, since it means
-                            -- that our clients records are broken
-                            -- (we're subscribed to something no
-                            -- client needs).
-                            error $ "subToClients: Nobody got the subscription " <>
-                                    "message for key " <> show k
-            other -> ztCliLog Warning $ "subToClients: wrong format: " <> show other
-
-    liftIO $ A.withAsync (heartbeatWorker cEnv) $ const $ do
-      let receiveBack = whileM (canReceive ztCliBack) $
-                        Z.receiveMulti ztCliBack >>= backendToClients
-      let receiveSub  = whileM (canReceive ztCliSub) $
-                        Z.receiveMulti ztCliSub >>= subToClients
+
+            -- Getting internal identifier of the node provided (if provided)
+            iIdM <- runMaybeT $ do
+                nodeId <- MaybeT $ pure nodeIdM
+                MaybeT $ HMap.lookup nodeId <$>
+                    atomically (readTVar (ztPeers ^. pivConnected))
+
+            if isJust nodeIdM && isNothing iIdM
+            then ztCliLog Warning $ "Coludn't send to peer, can't resolve: " <> show nodeIdM
+            else do
+                -- At this moment either both are nothing or both are just.
+
+                -- Choosing at random if not provided (both are nothing).
+                (iIdFinal :: Maybe ZTInternalId) <-
+                    maybe (fmap snd <$> choosePeer) (pure . Just) iIdM
```

Comment:
There is hell of a lot `let`-functions in code in `runBroker`. Can you make them `where`-functions instead, so the body will find itself near the function header?

Reply:
I can't because they use the contexts which I `view` in the beginning of `runBroker`.

Reply:
I recommend duplicating the reading statement to them.

I cannot fit this function on my mind at once.